packages/toolkit/src/query/core/buildInitiate.ts

```diff
@@ -276,20 +276,35 @@ Features like automatic cache collection, automatic refetching etc. will not be
         })
         const thunkResult = dispatch(thunk)
         middlewareWarning(getState)
-        const { requestId, abort, unwrap } = thunkResult
+
+        const { requestId, abort } = thunkResult
+
+        const aggregatePromise = Promise.all([
+          runningQueries[queryCacheKey],
+          thunkResult,
+        ]).then(() =>
+          (api.endpoints[endpointName] as ApiEndpointQuery<any, any>).select(
+            arg
+          )(getState())
+        )
+
         const statePromise: QueryActionCreatorResult<any> = Object.assign(
-          Promise.all([runningQueries[queryCacheKey], thunkResult]).then(() =>
-            (api.endpoints[endpointName] as ApiEndpointQuery<any, any>).select(
-              arg
-            )(getState())
-          ),
+          aggregatePromise,
           {
             arg,
             requestId,
             subscriptionOptions,
             queryCacheKey,
             abort,
-            unwrap,
+            unwrap() {
+              return aggregatePromise.then<any>((result) => {
+                if (result.isError) {
+                  throw result.error
+                }
+
+                return result.data
+              })
+            },
```

Comment:
The only thing that I think that could be interesting _(or confusing?)_ here is what happens regarding `unwrap` and `abort`. If abort is called, the initial request will be aborted, but a consumer will still be able to unwrap this. At the same time, I wouldn't want the abort to ignore the aggregate result, if that makes sense. I don't think there is any real issue with this implementation, but it _could be_ confusing. I don't have any better ideas or solutions for this though.

Reply:
`.abort` of the originally running promise should set the query into an error state, so I think this would result in an error here, which seems reasonable.

You bring up another problem though: `.abort()` on this would do nothing if it just piggy-backs on another promise that started earlier. But otoh, I guess that's okay, too - you would not want to abort something else by accident I guess?

Reply:
Can we make this function `async`? The .then usage here irritates me for some reason :sweat_smile: